### PR TITLE
Split predict into train/test

### PR DESCRIPTION
**Pull request by turian**
We still use the unlearned predictor, which we simply return from the train method.

However, now we do label mapping and properly return labels for multiclass and multilabel cases.

### CODE REVIEW

File: heareval/predictions/task_predictions.py [hunk 1]
```diff
@@ -68,18 +82,37 @@ def __init__(self, embedding_path: Path, split_name: str):
...
-        # return len(self.labels)
 
-    def __getitem__(self, idx):
-        return self.embedding_memmap[idx]
-        # return self.embedding_memmap[idx], self.labels[idx]
+    def __getitem__(self, idx) -> Tuple[torch.Tensor, torch.Tensor]:
+        x = self.embedding_memmap[idx]
+        y = [self.label_to_idx[str(label)] for label in self.labels[idx]]
+        if self.prediction_type == "multilabel":
+            # Lame special case
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:100]: Special case being a sample with zero labels?

File: heareval/predictions/task_predictions.py [hunk 2]
```diff
@@ -235,14 +273,35 @@ def task_predictions(
...
+    predictor: torch.nn.Module,
+    embedding_path: Path,
+    scene_embedding_size: int,
+    timestamp_embedding_size: int,
+):
+    metadata = json.load(embedding_path.joinpath("task_metadata.json").open())
+    label_vocab, nlabels = label_vocab_nlabels(embedding_path)
+
+    if metadata["embedding_type"] == "scene":
+        predictor = RandomProjectionPrediction(
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:289]: Don't we want to use the passed in predictor in this function?

File: heareval/predictions/task_predictions.py [hunk 3]
```diff
@@ -41,6 +42,7 @@ def __init__(self, nfeatures: int, nlabels: int, prediction_type: str):
             raise ValueError(f"Unknown prediction_type {prediction_type}")
 
     def forward(self, x: torch.Tensor):
+        x, y = x
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:45]: What's `y` for?

Comment (jorshi) [reply heareval/predictions/task_predictions.py:45]: Wait, so is the x that is passed in here a tuple that contains the embeddings and the ground truth? That's quite confusing, can we split that out before this?

File: heareval/predictions/task_predictions.py [hunk 4]
```diff
@@ -92,14 +125,15 @@ def create_events_from_prediction(
     converts the prediction tensor to a binary label based on the threshold value. Any
     events occurring at adjacent timestamps are considered to be part of the same event.
     This loops through and creates events for each label class. Disregards events that
-    are less than the min_duration.
+    are less than the min_duration milliseconds.
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:128]: nice

File: heareval/predictions/task_predictions.py (hunk 1 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:100]: yes

File: heareval/predictions/task_predictions.py (hunk 2 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:289]: No the passed in predictor was used to create *embeddings*. We learn our own predictor on top. But we afterwards will change this to single hidden layer.

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (turian) [reply heareval/predictions/task_predictions.py:45]: I fix this in a later PR

File: heareval/predictions/task_predictions.py [hunk 5]
```diff
@@ -35,28 +49,76 @@ def __init__(self, nfeatures: int, nlabels: int, prediction_type: str):
         torch.nn.init.normal_(self.projection.weight)
         if prediction_type == "multilabel":
             self.activation: torch.nn.Module = torch.nn.Sigmoid()
+            self.logitloss = torch.nn.BCEWithLogitsLoss()
```
Comment (khumairraj) [inline heareval/predictions/task_predictions.py:52]: I think this should be `self.logit_loss`

File: heareval/predictions/task_predictions.py [hunk 6]
```diff
@@ -35,28 +49,76 @@ def __init__(self, nfeatures: int, nlabels: int, prediction_type: str):
...
         elif prediction_type == "multiclass":
             self.activation = torch.nn.Softmax()
+            self.logit_loss = OneHotToCrossEntropyLoss()
```
Comment (khumairraj) [inline heareval/predictions/task_predictions.py:55]: Do we need `torch.nn.Softmax()` as activation? We already are using `torch.nn.CrossEntropyLoss()` in the `OneHotToCrossEntropyLoss()`, which does softmax before applying the cross entropy?

**Review by khumairraj — APPROVED** [review]

File: heareval/predictions/task_predictions.py (hunk 5 above)
Comment (turian) [reply heareval/predictions/task_predictions.py]: Done

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (turian) [reply heareval/predictions/task_predictions.py]: You can see now this is fixed

File: heareval/predictions/task_predictions.py (hunk 2 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:357]: I'm actually still a bit confused by this -- you pass in the predictor as a arg to this function and then overwrite it immediately? What's the point in passing in the predictor?

File: heareval/predictions/task_predictions.py [hunk 7]
```diff
@@ -215,44 +296,77 @@ def get_events_for_all_files(
...
         )
     elif metadata["embedding_type"] == "event":
-        predictor = RandomProjectionPrediction(
+        predictor = PredictionModel(
             timestamp_embedding_size, nlabels, metadata["prediction_type"]
         )
     else:
         raise ValueError(f"embedding_type {metadata['embedding_type']} unknown.")
+    # train on CPU
+    # TODO: FIXME
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:339]: What's broken? Can you expand on this comment a bit?

File: heareval/evaluation/task_evaluation.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+"""
+Contains a set of various evaluation scores. Runs evaluation on the test predictions
+for predictions on a HEAR task.
+Would it make sense to move all the score functions over to a file called scores?
+
+TODO: Fix AUC
```
Comment (jorshi) [inline heareval/evaluation/task_evaluation.py:7]: Update docstring

**Review by jorshi — APPROVED** [review]
